dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AccountController.java

```diff
@@ -222,6 +222,225 @@ public WebMessage restoreAccount( @RequestParam String token, @RequestParam Stri
+        }
+
+        if ( password == null )
+        {
+            return badRequest( "Password is not specified" );
+        }
+
+        PasswordValidationResult result = passwordValidationService.validate( credentialsInfo );
+
+        if ( !result.isValid() )
+        {
+            return badRequest( result.getErrorMessage() );
+        }
+
+        if ( email == null || !ValidationUtils.emailIsValid( email ) )
+        {
+            return badRequest( "Email is not specified or invalid" );
+        }
+
+        if ( phoneNumber == null || phoneNumber.trim().length() > MAX_PHONE_NO_LENGTH )
+        {
+            return badRequest( "Phone number is not specified or invalid" );
+        }
+
+        if ( employer == null || employer.trim().length() > MAX_LENGTH )
+        {
+            return badRequest( "Employer is not specified or invalid" );
+        }
+
+        if ( !systemSettingManager.selfRegistrationNoRecaptcha() )
+        {
+            if ( recapResponse == null )
+            {
+                return badRequest( "Please verify that you are not a robot" );
+            }
+
+            // ---------------------------------------------------------------------
+            // Check result from API, return 500 if validation failed
+            // ---------------------------------------------------------------------
+
+            RecaptchaResponse recaptchaResponse = securityService
+                .verifyRecaptcha( recapResponse, request.getRemoteAddr() );
+
+            if ( !recaptchaResponse.success() )
+            {
+                log.warn( "Recaptcha validation failed: " + recaptchaResponse.getErrorCodes() );
+                return badRequest( "Recaptcha validation failed: " + recaptchaResponse.getErrorCodes() );
+            }
+        }
+
+        // ---------------------------------------------------------------------
+        // Create and save user, return 201
+        // ---------------------------------------------------------------------
+
+        if ( invitedByEmail )
+        {
+            boolean restored = securityService.restore( user, restoreToken, password, RestoreType.INVITE );
+
+            if ( !restored )
+            {
+                log.info( "Invite restore failed for: " + inviteUsername );
+
+                return badRequest( "Unable to create invited user account" );
+            }
+
+            user = new User();
+            user.setFirstName( firstName );
+            user.setSurname( surname );
+            user.setEmail( email );
+            user.setPhoneNumber( phoneNumber );
+            user.setEmployer( employer );
+
+            if ( canChooseUsername )
+            {
+                user.setUsername( username );
+            }
+            else
+            {
+                username = user.getUsername();
+            }
+
+            userService.encodeAndSetPassword( user, password );
+
+            userService.updateUser( user );
```

Comment:
Should this be `userService.addUser( user )` ? since we have `user = new User();` above